FastJNI support for AES-CBC

ashbm5 opened the pull request:
New FastJNI support for AES-CBC algorithms implemented by JCE providers.

It includes the ability to release VMAccess mid-way through the processing of very large input data. The way this is done is by detecting the operation type and breaking it up. A single large update operation will be converted into a sequence of smaller update operations where the VAccess lock is released between operation invocation. Similarly, a large final operation will be converted to a sequence of update calls and a single final invocation at the end of the operation.

This support is targeted at JCE providers that use an external native library via JNI in order to provide cryptography routines. In order to leverage this support, there are two requirements on the provider:
1) The JNI signature needs to be of a certain form as specified in the code.
2) The native library that is used by the JCE provider must implement specific routines that perform the update and final operations for AES-CBC.

Note: dis-contiguous arrays are handled by allocating a contiguous array of the same size and having the JCE operate on that new array. There is no support for releasing VMAccess lock during operations on very large dis-contiguous arrays.

Signed-off-by: Alon Sh alonsh@ca.ibm.com

ashbm5 commented:
FYI @gacholio @vijaysun-omr @joransiu @zl-wang

gacholio commented on runtime/vm/jvminit.c:
```diff
@@ -719,6 +719,14 @@ freeJavaVM(J9JavaVM * vm)
 
+#if !defined(J9ZOS390)
+	if (NULL != vm->jniCryptoFunctions) {
+		if (vm->jniCryptoLibrary) {
```
Unnecessary - if you've found functions, you've opened the library.

gacholio commented on runtime/oti/j9nonbuilder.h:
```diff
@@ -5171,6 +5171,9 @@ typedef struct J9JavaVM {
 	UDATA modulePointerOffset;
 	omrthread_monitor_t jniCriticalLock;
+	omrthread_monitor_t jniCryptoLibLock;
+        void* jniCryptoFunctions;
```
Spacing.

gacholio commented on runtime/vm/CMakeLists.txt:
```diff
@@ -44,6 +44,7 @@ j9vm_add_library(j9vm SHARED
 	exceptiondescribe.c
 	exceptionsupport.c
 	extendedMessageNPE.cpp
+        FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp
```
Spacing.

gacholio commented on runtime/vm/vmthinit.c:
```diff
@@ -55,7 +55,7 @@ UDATA initializeVMThreading(J9JavaVM *vm)
 		omrthread_monitor_init_with_name(&vm->fieldIndexMutex, 0, "Field Index Hashtable Mutex") ||
 		omrthread_monitor_init_with_name(&vm->jniCriticalLock, 0, "JNI critical region mutex") ||
-
+		omrthread_monitor_init_with_name(&vm->jniCryptoLibLock, 0, "JNI Cryptography mutex") ||
```
Monitor should be destroyed in `terminateVMThreading`.

ashbm5 commented:
@gacholio updated

gacholio commented:
@ashbm5 You don't seem to have signed the ECA.

ashbm5 commented:
@gacholio done

gacholio commented on runtime/vm/jvminit.c:
```diff
@@ -719,6 +719,12 @@ freeJavaVM(J9JavaVM * vm)
 		vm->jniGlobalReferences = NULL;
 	}
 
+#if !defined(J9ZOS390)
```
Why the ifdef? If anything, I'd expect to only do this on ZOS, not everywhere but. The fields aren't ifdeffed.

ashbm5 replied:
this is due to the comment starting at line 746.

```
                /* zOS: Do not close any of the DLLs. This is necessary 
                 * because we do not know for sure whether all the threads 
                 * that may depend on these DLLs have terminated.
                 *
                 * Note that this solution is also suggested in the 
                 * 'z/OS XL C/C++ Programming Guide' V1R13 (Chapter 21 
                 * under the heading 'DLL Restrictions').
                 */
#if !defined(J9ZOS390)
                closeAllDLLs(vm);
                /* Remember the file descriptor of the trace DLL. This has to be closed later than other DLLs. */
                traceLoadInfo = FIND_DLL_TABLE_ENTRY(J9_RAS_TRACE_DLL_NAME);
                if (NULL != traceLoadInfo) {
                        traceDescriptor = traceLoadInfo->descriptor;
                }    
#endif /* !defined(J9ZOS390) */
```

gacholio replied:
Good catch - please move the close code into `closeAllDLLs`

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+/*******************************************************************************
+ * Copyright (c) 2001, 2020 IBM Corp. and others
```
Should be 2020, 2020.

zl-wang replied:
Nah ... I just completed the required Developer Legal Education.  It should be 2001, 2020.   The first number is the year in which the first time commit or publish.  The second number is the year in which the latest modification is.

zl-wang replied:
is this completely new code?  then, it should be 2020,2020

zl-wang replied:
if this is a modification to some J9 code, it could be 2001,2020.

gacholio replied:
This is a new file.

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+ * Need to synchronize this operation so that threads won't
+ * use the routines too early
+ * */
+void loadCryptoLib(J9VMThread *currentThread) {
```
VM standard for function declarations is:
```
void
loadCryptoLib(J9VMThread *currentThread)
{
```

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+ * */
+void loadCryptoLib(J9VMThread *currentThread) {
+
+	if (currentThread->javaVM->jniCryptoFunctions == NULL) {
```
Constant test should be reversed - `NULL == variable`

gacholio replied:
As a rule, this only applies to equal/notequal tests - `a < 5` is certainly better than `5 >= a`

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+			const char * decryptUpdate_name = "CIPHER_decryptUpdate_internal";
+			const char * zKMC_name = "CIPHER_zKMC_internal";
+	
+			j9sl_open_shared_library((char*)cryptolib_name, cryptoLibrary, J9PORT_SLOPEN_DECORATE);
```
Error checks required for all port library calls.

ashbm5 replied:
updated the error handling as per feedback.

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+#ifdef CRYPTO_DEBUG
+			printf("** %d bytes out of %d\n", bytesToOp, totalBytes);
+#endif
+			currentBytes = (bytesToOp-OP_SIZE_LIMIT >= 0) ? OP_SIZE_LIMIT : bytesToOp;
```
Math should be spaced `a - b` and brackets should be used - `(a - b) >= 0`

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+	jint ret = 0;
+	I_32 outputArraySize = J9INDEXABLEOBJECT_SIZE(currentThread, output);
+
+	PORT_ACCESS_FROM_VMC(currentThread);
```
This should be treated as a variable declaration.

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+			CIPHER_t* CIPHER_encryptFinal;
+			CIPHER_t* CIPHER_decryptFinal;
+			z_kmc_native_t* z_kmc_native;
+			const char * cryptolib_name = "jgskit";
```
I don't see the point of declaring these rather than just using them directly in the port library calls.

ashbm5 replied:
without declaring those. For example, if the code was

``j9sl_open_shared_library("jgskit", cryptoLibrary, J9PORT_SLOPEN_DECORATE);``

The compiler will generate an error such as:

```FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp: In function 'void loadCryptoLib(J9VMThread*)':
../oti/j9port_generated.h:553:180: error: ISO C++ forbids converting a string constant to 'char*' [-Werror=write-strings]
 #define j9sl_open_shared_library(param1,param2,param3) OMRPORT_FROM_J9PORT(privatePortLibrary)->sl_open_shared_library(OMRPORT_FROM_J9PORT(privatePortLibrary),param1,param2,param3)
```

gacholio replied:
Fair enough - the correct solution would be to consitfy the port library, but that's not going to happen here.

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+			cryptoPointers_t* cryptFuncs = (cryptoPointers_t*)j9mem_allocate_memory(sizeof(cryptoPointers_t), J9MEM_CATEGORY_VM_JCL);
+			UDATA* cryptoLibrary = &(currentThread->javaVM->jniCryptoLibrary);
+	
+			CIPHER_t* CIPHER_encryptFinal;
```
(almost) All variables should be initialized on declaration.

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+		omrthread_monitor_enter(mutex);
+		
+		if (currentThread->javaVM->jniCryptoFunctions == NULL) {
+	
```
Line spacing seems a bit random - normal standard is to declare variables immediately, possibly followed by blank line before the code begins.

gacholio commented on runtime/vm/FastJNI_com_ibm_crypto_plus_provider_icc_NativeInterface.cpp:
```diff
@@ -0,0 +1,415 @@
+
+	jint totalBytes = inputLen;
+
+	if (totalBytes > 0) {
```
No need for this check.

ashbm5 replied:
if the value is negative, no need to handle it in vm code - the JCL code will take care of throwing exceptions etc as needed. negative value doesn't impact the functionality of the vm code - a single shot call will be invoked with negative bytes and then return.

gacholio replied:
OK, then the check should be removed (and comment this fact in the function doc).

ashbm5 replied:
If totalBytes is zero, the JCL is indicating that the input array is null and therefore I skip the line and leave inputAddress at NULL. 
``inputAddress = J9JAVAARRAY_EA(currentThread, in, inputIndex, U_8);``

gacholio replied:
Given that:

- the length will essentially never be 0
- the array macro will work fine with a 0-sized array
- 0-sized arrays are handled by the subsequent code

The check is not needed.

gacholio replied:
It's still worth commenting on the ensured validity of the input arguments.

ashbm5 replied:
will the macro work if ``in`` is a null in java?

gacholio replied:
No, it will not work with NULL. Perhaps check for that instead of the size, assuming that both will be 0 for the size computation case.

gacholio commented:
I'm not sure exactly what you want to do, but there needs to be some fallback for if the library cannot be loaded.

gacholio commented:
Is this feature specific to ZOS?

ashbm5 commented:
no, its cross platform for both jdk8 and jdk11

gacholio commented:
Is the native library already included in *every* build (open and closed)?

ashbm5 commented:
The native library is optional for open and is always included in closed builds. The JNI signatures that are being recognized are internal and specific to the native library implementation so when the native library is not enabled in the open builds, the fastjni code will not be triggered.

gacholio commented:
As long as the library is included in every build JDK8 or higher, then the assert is probably OK. You will still need to check for memory allocation failure in the natives themselves and throw OOM (so the natives will need to be tagged appropriately - remove the "no exception" and "no special teardown" flags).

gacholio commented:
More general comments - we prefer `#if defined` rather than `#ifdef`, and `#else` and `#endif` should be commented.

ashbm5 commented:
updated the error handling to throw OOM exceptions when fail on allocation (two places, on lib. load and when handling dis-contiguous arrays). also updated to vm assert when failing to load crypto. library. finally updated the ifdefs per review and tried to make the code a bit easier to read by combining debug ifdefs when possible.

gacholio commented on runtime/vm/jvminit.c:
```diff
@@ -4066,6 +4067,13 @@ static void closeAllDLLs(J9JavaVM* vm) {
+	if (NULL != vm->jniCryptoFunctions) {
+		const char * cryptolib_name = "jgskit";
+
+		j9sl_close_shared_library(vm->jniCryptoLibrary);
```
Should also free (and NULL) the allocated function table.

ashbm5 replied:
updated

mstoodle commented:
If this change is only to support optimization of a closed source library, then I think this pull request is better applied by the owner of that closed source binary in a downstream fork of OpenJ9 rather than in the upstream.

If that closed source library is open sourced and can be included in OpenJ9 builds, then we can revisit accepting this change at OpenJ9.

I recommend this pull request be closed, but I'll leave it open for comment for the time being.
